### PR TITLE
Initial tracing framework via CoroutineContext

### PR DESCRIPTION
**Pull request by ianbotsf**
## Issue \#

#677 

## Description of changes

Counterproposal to #707.

**Companion PR**: [aws-sdk-kotlin#731](https://github.com/awslabs/aws-sdk-kotlin/pull/731)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: runtime/tracing/tracing-core/common/src/aws/smithy/kotlin/runtime/tracing/CoroutineContextUtils.kt [hunk 1]
```diff
@@ -4,224 +4,211 @@
...
  */
-public inline fun <R> ExecutionContext.withChildSpan(id: String, block: () -> R): R {
-    pushChildTraceSpan(id)
-    try {
-        return block()
+public suspend inline fun <R> CoroutineContext.withChildTraceSpan(
+    id: String,
+    crossinline block: suspend CoroutineScope.() -> R,
+): R {
+    val existingSpan = checkNotNull(get(TraceSpanContextElement)?.traceSpan) { "Missing an active trace span" }
```
Comment (aajtodd) [inline runtime/tracing/tracing-core/common/src/aws/smithy/kotlin/runtime/tracing/CoroutineContextUtils.kt:38]: Should this error or should it just create a no-op trace span?

That seems to be what open telemetry does FWIW:

[Span.current()](https://javadoc.io/static/io.opentelemetry/opentelemetry-api/1.19.0/io/opentelemetry/api/trace/Span.html#current())
> Returns the [Span](https://javadoc.io/static/io.opentelemetry/opentelemetry-api/1.19.0/io/opentelemetry/api/trace/Span.html) from the current Context, falling back to a default, no-op [Span](https://javadoc.io/static/io.opentelemetry/opentelemetry-api/1.19.0/io/opentelemetry/api/trace/Span.html) if there is no span in the current context.

I think that most library code probably shouldn't care. If it's not instrumented with a parent span somewhere then it's a no-op seems fine to me. Only when the calling environment provides the span (e.g. when we generate operation setup) will it be instrumented. 

e.g.:

```kotlin
public suspend inline fun <R> span(
    id: String,
    crossinline block: suspend () -> R
): R {
    val currentSpan = // get current or no-op span
    val childSpan = currentSpan.child(id)
    return try {
        withContext(TraceSpanContext(childSpan) {
            block()
        }
     } finally {
         childSpan.close()
     }
}
```

Use:

```kotlin
suspend fun foo() = span("foo") { ... }
```

File: smithy-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpProtocolClientGeneratorTest.kt [hunk 2]
```diff
@@ -196,7 +196,7 @@ class HttpProtocolClientGeneratorTest {
         }
         op.install(MockMiddleware(configurationField1 = "testing"))
         val rootSpan = config.tracer.createRootSpan("GetFooStreamingInputNoOutput-${'$'}{op.context.sdkRequestId}")
-        return op.context.withRootSpan(rootSpan) {
+        return coroutineContext.withRootTraceSpan(rootSpan) {
```
Comment (aajtodd) [inline smithy-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpProtocolClientGeneratorTest.kt:199]: This seems like we don't need a wrapper necessarily for a root span, wouldn't this suffice:

```kotlin
    return withContext(TraceSpanContext(rootSpan)) { ... }
```

File: runtime/tracing/tracing-core/common/src/aws/smithy/kotlin/runtime/tracing/CoroutineContextUtils.kt [hunk 3]
```diff
@@ -4,224 +4,211 @@
...
+        }
     } finally {
-        popChildTraceSpan()
+        childSpan.close()
     }
 }
 
 @InternalApi
-public inline fun <R> ExecutionContext.withRootSpan(rootSpan: TraceSpan, block: () -> R): R =
+public suspend inline fun <R> CoroutineContext.withRootTraceSpan(
```
[the remark this answers is not in the document]
Comment (ianbotsf) [reply runtime/tracing/tracing-core/common/src/aws/smithy/kotlin/runtime/tracing/CoroutineContextUtils.kt:50]: There were two reasons I chose to separate `withRootTraceSpan` from `withChildTraceSpan`:
* The child variant accepts a string id and passes that to the active span's `child` method, simplifying the call site. The root variant must, by necessity, accept a full `TraceSpan`.
* The child variant is expected to be invoked by users in at least some scenarios. The root variant is only intended for internally setting up the initial, root span (and is accordingly marked `@InternalApi`).

File: runtime/tracing/tracing-core/common/src/aws/smithy/kotlin/runtime/tracing/CoroutineContextUtils.kt [hunk 4]
```diff
@@ -4,224 +4,211 @@
...
- * @param block The block of code to execute. Once the block is complete, the previous span will be restored.
+ * Runs a block of code within the context of a child [TraceSpan]. This call pushes the new child trace span onto the
+ * coroutine context before executing [block] and restores the parent span after the block completes (whether
+ * successfully or exceptionally).
  */
-public inline fun <R> ExecutionContext.withChildSpan(id: String, block: () -> R): R {
-    pushChildTraceSpan(id)
-    try {
-        return block()
+public suspend inline fun <R> CoroutineContext.withChildTraceSpan(
```
[the remark this answers is not in the document]
Comment (ianbotsf) [reply runtime/tracing/tracing-core/common/src/aws/smithy/kotlin/runtime/tracing/CoroutineContextUtils.kt:34]: Oh good point, that's true for `withChildTraceSpan` and `withRootTraceSpan`. That's not as true for `log`/`info`/`error`, etc. Those methods currently don't need to be `suspend`, they merely need a way to get the active `TraceSpan`. Making them `suspend` will add additional state machine overhead that feels unwarranted to me, although perhaps there's room for debate. And I don't believe that `suspend` could be applied to the `traceSpan` property either.

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/RuntimeTypes.kt [hunk 5]
```diff
@@ -230,10 +230,14 @@ object RuntimeTypes {
             val KotlinLoggingTraceProbe = runtimeSymbol("KotlinLoggingTraceProbe", KotlinDependency.CORE, "tracing")
             val TraceProbe = runtimeSymbol("TraceProbe", KotlinDependency.CORE, "tracing")
             val Tracer = runtimeSymbol("Tracer", KotlinDependency.CORE, "tracing")
-            val withRootSpan = runtimeSymbol("withRootSpan", KotlinDependency.CORE, "tracing")
+            val withRootTraceSpan = runtimeSymbol("withRootTraceSpan", KotlinDependency.CORE, "tracing")
         }
     }
 
+    object KotlinCoroutines {
```
[the remark this answers is not in the document]
Comment (ianbotsf) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/RuntimeTypes.kt:237]: Yes but `kotlin.coroutines` is a different namespace than `kotlinx.coroutines`. Would you like prefer these to use the same `KotlinxCoroutines` object regardless?

File: smithy-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpProtocolClientGeneratorTest.kt (hunk 2 above)
Comment (ianbotsf) [reply smithy-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpProtocolClientGeneratorTest.kt:199]: The `withRootTraceSpan` method also verifies that no other trace span is active and then closes the root when the lambda is finished.

File: runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/engine/CoroutineUtils.kt [hunk 6]
```diff
@@ -20,7 +22,10 @@ import kotlin.coroutines.coroutineContext
 internal fun HttpClientEngine.createCallContext(outerContext: CoroutineContext): CoroutineContext {
     // attach request to engine (will ensure that shutdown only is invoked after all in-flight requests complete)
     val requestJob = Job(coroutineContext[Job])
-    val reqContext = coroutineContext + requestJob + CoroutineName("request-context")
+    val reqContext = coroutineContext +
+        requestJob +
+        CoroutineName("request-context") +
+        TraceSpanContextElement(outerContext.traceSpan)
```
[the remark this answers is not in the document]
Comment (ianbotsf) [reply runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/engine/CoroutineUtils.kt:28]: 19 days later I cannot remember which precise issue I had described to the team which might be solved by this change. I can say that _without_ this change, none of the engine logging would have a span which is currently an error.

The question of whether a span must exist is closer to a tenet decision...moving that discussion offline.

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/RuntimeTypes.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (aajtodd) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/RuntimeTypes.kt:237]: Oh nevermind, disregard

**Comment by ianbotsf** [comment]
Closing in favor of cleaner merge in #737.

**Pull request closed without merging**